### ISSUE TITLE
Resource and Data elements a'la XAML.

### ISSUE THREAD

**Issue by kevmeister68** [aurelia/templating#317]
One of the really neat features of XAML is the fact that every element in XAML is ultimately the creator of an object first-and-foremost, and it just so happens that some of those objects are involved in presentation. Of course HTML cannot be XAML because it is first and foremost a presentation format, but there are various things Aurelia could do in its templating engine to assist in bridging the gap between the two.

I won't pretend that this is a fully designed idea, in the sense I don't know what impediments stand in the way of achieving it.

So, what's missing:

#1 

XAML Example: `<my-custom-class x:Name="myPropertyName" attrib1="blah" attrib2="blah"></my-custom-class>`

Aurelia example: `<my-custom-class ref="myPropertyName" attrib1="blah" attrib2="blah"></my-custom-class>`

At this level, XAML and Aurelia are fairly close, *if* you define the my-custom-class in Aurelia with the @noView decorator.

Here's where XAML gets useful:
```xaml
<my-custom-class x:Name="myPropertyName" attrib1="blah">
   <my-custom-class.attrib2>
      <another-custom-class name="Kevin"></another-custom-class>
   </my-custom-class.attrib2>
</my-custom-class>
```
The above example is not achievable in Aurelia in a general sense. Notice the inner element <my-custom-class.attrib2> makes it known that the content of this element is intended for the "attrib2" property of my-custom-class' object (view-model).

The XAML syntax does not just allow single objects to be assigned, it also supports collections of objects. XAML has the benefit (compared to JS) of having Reflection available to it), but it knows when it is providing child items for a collection object, and behaves accordingly.

#2 

Probably less important.

A general place to place collections of resource objects that are declared in markup. The XAML example:
```xaml
<my-page>
   <my-page.resources>
      <my-custom-class x:Key="barney"></my-custom-class>
   <my-page.resources>
</my-page>
```
From memory, XAML has x:Key and x:Name because X:Name produced class property, and x:Key did not. In case it is not obvious, the above is really just drawing on #1's ability to populate a collection with objects.

I have noticed at least one example (https://github.com/aurelia/templating/issues/162) where people have requested an ability for markup to represent data, and the above provides that capability also.

Thanks for reading :-)

**Comment by EisenbergEffect** [comment]
I can't actually see the markup examples here. so it's a bit hard to follow...

**Comment by jdanyow** [comment]
cleaned up the markup... @kevmeister68 would this work for you:

xaml:
```xaml
<my-custom-class attrib1="blah">
   <my-custom-class.attrib2>
      <another-custom-class name="Kevin"></another-custom-class>
   </my-custom-class.attrib2>
</my-custom-class>
```

html/aurelia:
```html
<my-custom-class attrib1="blah" attrib2.bind="{ name: 'Keven' }">
</my-custom-class>
```

**Comment by kevmeister68** [comment]
Thanks for fixing the markup, I got sidetracked and didn't check what I posted :-)

I guess the first point I'd make is that I'm not looking to solve or workaround a current problem I have, and am instead making a suggestion to achieve a usability improvement when you want to use markup as a vehicle to create object structures.

@jdanyow the suggestion you have is not ideal because it involves a plain-old-object, not a class. The nice thing about XAML (as experienced through Silverlight) is that object creation is pretty much "equivalent" whether you do it in code or in markup.

On the other hand, philosophically XAML was always intended to basically just represent an object hierarchy (or many of them), and the fact that some of those objects were visual objects (or not) almost seemed a secondary consideration. No matter how much I pretend, HTML is not XAML and so philosophically this might be a "sticking point" in terms of trying to make HTML behave more like XAML in that regard. Another potential sticking point is whether creating objects in the "view" violates what the role of the "view" actually is (on the other hand, Silverlight was a MVVM architecture and had no problem offering this capability).

Having said that, though, creation of class-backed objects in markup with an appropriately capable syntax can often be more productive than jumping into code.

**Comment by EisenbergEffect** [comment]
I'm going to close this for now since I don't think it's possible to achieve exactly what Xaml does since HTML has a DOM while Xaml is just an object hierarchy. It is possible to create objects if you specify they have @noView() and give them @bindable properties. In the future, I hope to remove the need for @bindable, but that is pending some future JS language reflection features for fields. So, as time passes, this could get easier.
